src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -140,264 +35,133 @@ EFeatureType RandomFourierGaussPreproc::get_feature_type() {
-	}
-
-	dim_input_space = dim;
-
-}
-
-bool RandomFourierGaussPreproc::test_rfinited() const {
-
-	if ((dim_feature_space ==  cur_dim_feature_space)
-			&& (dim_input_space > 0) && (dim_feature_space > 0)) {
-		if ((dim_input_space == cur_dim_input_space)&&(Math::abs(kernelwidth-cur_kernelwidth)<1e-5)) {
+void RandomFourierGaussPreproc::init()
+{
+	m_kernelwidth = 1;
+	m_dims = 1000;
 
-			// already inited
-			return true;
-		} else {
-			return false;
-		}
-	}
+	m_offset = SGVector<float64_t>();
+	m_basis = SGMatrix<float64_t>();
 
-	return false;
+	SG_ADD(&m_dims, "m_dims", "Dimensionality of the output feature space.");
+	SG_ADD(&m_kernelwidth, "m_kernelwidth", "Kernel width.",
+	    ParameterProperties::HYPER);
+	SG_ADD(&m_basis, "m_basis", "Matrix of basis vectors / 1 vector per column");
+	SG_ADD(&m_offset, "m_offset", "offset vector");
 }
 
-bool RandomFourierGaussPreproc::init_randomcoefficients() {
-	if (dim_feature_space <= 0) {
-		throw ShogunException(
-				"bool CRandomFourierGaussPreproc::init_randomcoefficients(): dim_feature_space<=0 is not allowed\n");
-	}
-	if (dim_input_space <= 0) {
-		throw ShogunException(
-				"bool CRandomFourierGaussPreproc::init_randomcoefficients(): dim_input_space<=0 is not allowed\n");
-	}
-
-	if (test_rfinited()) {
-		return false;
-	}
-
-
-	io::info("initializing randomcoefficients ");
-
-	float64_t pi = 3.14159265;
+void RandomFourierGaussPreproc::generate_coeff(int32_t dim_input_space) {
 
+	NormalDistribution<float64_t> normal_dist;
+	UniformRealDistribution<float64_t> uniform_real_dist(0.0, 2 * Math::PI);
 
-	SG_FREE(randomcoeff_multiplicative);
-	randomcoeff_multiplicative=NULL;
-	SG_FREE(randomcoeff_additive);
-	randomcoeff_additive=NULL;
+	m_basis = SGMatrix<float64_t>(m_dims,dim_input_space);
+	m_offset = SGVector<float64_t>(m_dims);
 
-
-	cur_dim_feature_space=dim_feature_space;
-	randomcoeff_additive=SG_MALLOC(float64_t, cur_dim_feature_space);
-	cur_dim_input_space = dim_input_space;
-	randomcoeff_multiplicative=SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
-
-	cur_kernelwidth=kernelwidth;
-
-	random::fill_array(
-	    randomcoeff_additive, randomcoeff_additive + cur_dim_feature_space, 0.0,
-	    2 * pi, m_prng);
-	
-	UniformRealDistribution<float64_t> uniform_real_dist(-1.0, 1.0);
-	for (int32_t  i = 0; i < cur_dim_feature_space; ++i) {
-		for (int32_t k = 0; k < cur_dim_input_space; ++k) {
-			float64_t x1,x2;
-			float64_t s = 2;
-			while ((s >= 1) ) {
-				// Marsaglia polar for gaussian
-				x1 = uniform_real_dist(m_prng);
-				x2 = uniform_real_dist(m_prng);
-				s=x1*x1+x2*x2;
-			}
-
-			// =  x1/std::sqrt(val)* std::sqrt(-2*std::log(val));
-			randomcoeff_multiplicative[i * cur_dim_input_space + k] =
-			    x1 * std::sqrt(-2 * std::log(s) / s) / kernelwidth;
+	for(auto row : range(m_dims))
+	{	
+		for(auto col : range(dim_input_space))
+		{
+			m_basis(row,col)= std::sqrt((float64_t)1 / m_kernelwidth) *
```

Comment:
could you move this to a method "sample_spectral_density". This will allow us to refactor things in terms of different kernels later

Reply:
A protected virtual method, I'm sure you agree?. We could define a superclass which encompasses all explicit feature mapping inheriting from DensePreprocessor. We could then put sample_spectral_density protected virtual pure in that class.

Reply:
yeah, that would be a next step, but let's not do too many things are once

Reply:
i.e. just moving the code to a method is fine, and later we can pick up on that